File: packages/go/libsql/crud_test.go
package libsql_test

import (
	"context"
	"database/sql"
	"path/filepath"
	"testing"

	"github.com/b4moss/crudian/go/crudian"
	libsqlcrud "github.com/b4moss/crudian/go/libsql"
	_ "github.com/tursodatabase/libsql-client-go/libsql"
	_ "modernc.org/sqlite" // required by libsql-client-go for file:// URLs
)

func openDB(t *testing.T) *sql.DB {
	t.Helper()
	path, err := filepath.Abs(filepath.Join(t.TempDir(), "test.db"))
	if err != nil {
		t.Fatalf("abs: %v", err)
	}
	// libsql file URLs must be file:///...; companion sqlite driver required.
	db, err := sql.Open("libsql", "file://"+filepath.ToSlash(path))
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })
	if err := db.Ping(); err != nil {
		t.Fatalf("ping: %v", err)
	}
	if _, err := db.Exec(`CREATE TABLE items (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		name TEXT NOT NULL,
		score INTEGER NOT NULL,
		note TEXT
	)`); err != nil {
		t.Fatalf("migrate: %v", err)
	}
	return db
}

func mustCreateCrud(t *testing.T, db *sql.DB) *libsqlcrud.Crud {
	t.Helper()
	crud, err := libsqlcrud.CreateCrud(db)
	if err != nil {
		t.Fatalf("CreateCrud: %v", err)
	}
	return crud
}

func TestCreateCrud(t *testing.T) {
	db := openDB(t)
	crud := mustCreateCrud(t, db)
	if crud.DB != db {
		t.Fatal("expected same *sql.DB")
	}
	if _, err := libsqlcrud.CreateCrud(nil); err == nil {
		t.Fatal("expected nil db error")
	}
}

func TestCoreCrud(t *testing.T) {
	ctx := context.Background()
	crud := mustCreateCrud(t, openDB(t))

	created, err := crud.Create(ctx, "items", crudian.Row{"name": "alice", "score": 10})
	if err != nil {
		t.Fatalf("create: %v", err)
	}
	if toInt64(created["id"]) == 0 || created["name"] != "alice" {
		t.Fatalf("create row: %+v", created)
	}
	id := created["id"]

	got, err := crud.Read(ctx, "items", crudian.ReadQuery{Where: crudian.Where().Eq("id", id)})
	if err != nil || got == nil || got["name"] != "alice" {
		t.Fatalf("read: %+v err=%v", got, err)
	}

	proj, err := crud.Read(ctx, "items", crudian.ReadQuery{
		Columns: []string{"name"},
		Where:   crudian.Where().Eq("id", id),
	})
	if err != nil || proj == nil || proj["name"] != "alice" {
		t.Fatalf("columns: %+v err=%v", proj, err)
	}
	if _, ok := proj["score"]; ok {
		t.Fatalf("score should be projected out: %+v", proj)
	}

	miss, err := crud.Read(ctx, "items", crudian.ReadQuery{Where: crudian.Where().Eq("id", 99999)})
	if err != nil || miss != nil {
		t.Fatalf("miss: %+v err=%v", miss, err)
	}

	updated, err := crud.Update(ctx, "items", crudian.Row{"score": 11}, crudian.UpdateQuery{Where: crudian.Where().Eq("id", id)})
	if err != nil || updated == nil || toInt64(updated["score"]) != 11 || updated["name"] != "alice" {
		t.Fatalf("update: %+v err=%v", updated, err)
	}

	none, err := crud.Update(ctx, "items", crudian.Row{"score": 1}, crudian.UpdateQuery{Where: crudian.Where().Eq("id", 99999)})
	if err != nil || none != nil {
		t.Fatalf("update miss: %+v err=%v", none, err)
	}

	n, err := crud.Delete(ctx, "items", crudian.DeleteQuery{Where: crudian.Where().Eq("id", id)})
	if err != nil || n != 1 {
		t.Fatalf("delete: %d %v", n, err)
	}
	z, err := crud.Delete(ctx, "items", crudian.DeleteQuery{Where: crudian.Where().Eq("id", id)})
	if err != nil || z != 0 {
		t.Fatalf("delete zero: %d %v", z, err)
	}

	if _, err := crud.Create(ctx, "", crudian.Row{"name": "x", "score": 1}); err == nil {
		t.Fatal("empty table should fail")
	}
}

func TestSearchListCount(t *testing.T) {
	ctx := context.Background()
	crud := mustCreateCrud(t, openDB(t))

	for _, name := range []string{"a", "b", "c"} {
		if _, err := crud.Create(ctx, "items", crudian.Row{"name": name, "score": 1}); err != nil {
			t.Fatal(err)
		}
	}

	page, err := crud.Search(ctx, "items", crudian.SearchQuery{Limit: 2})
	if err != nil {
		t.Fatalf("search: %v", err)
	}
	if len(page.Items) != 2 || !page.HasMore || page.NextCursor == nil || page.Total != 3 {
		t.Fatalf("page1: %+v", page)
	}
	page2, err := crud.Search(ctx, "items", crudian.SearchQuery{Limit: 2, Cursor: page.NextCursor})
	if err != nil || len(page2.Items) != 1 || page2.HasMore || page2.Total != 3 {
		t.Fatalf("page2: %+v %v", page2, err)
	}

	list, err := crud.List(ctx, "items", crudian.SearchQuery{Limit: 2})
	if err != nil || list.Total != page.Total || len(list.Items) != len(page.Items) {
		t.Fatalf("list alias: %+v vs %+v", list, page)
	}

	found, err := crud.Search(ctx, "items", crudian.SearchQuery{
		Where: crudian.Where().Eq("name", "b"),
		Limit: 10,
	})
	if err != nil || len(found.Items) != 1 || found.Total != 1 {
		t.Fatalf("eq: %+v %v", found, err)
	}

	or, err := crud.Search(ctx, "items", crudian.SearchQuery{
		Where: crudian.Where().Eq("name", "a").Or(crudian.Where().Eq("name", "c")),
		Limit: 10,
	})
	if err != nil || or.Total != 2 {
		t.Fatalf("or: %+v %v", or, err)
	}

	in, err := crud.Search(ctx, "items", crudian.SearchQuery{
		Where: crudian.Where().In("name", []any{"a", "b"}),
		Limit: 10,
	})
	if err != nil || in.Total != 2 {
		t.Fatalf("in: %+v %v", in, err)
	}

	like, err := crud.Search(ctx, "items", crudian.SearchQuery{
		Where: crudian.Where().Like("name", "a%"),
		Limit: 10,
	})
	if err != nil || like.Total != 1 {
		t.Fatalf("like: %+v %v", like, err)
	}

	if _, err := crud.Create(ctx, "items", crudian.Row{"name": "n", "score": 0}); err != nil {
		t.Fatal(err)
	}
	if _, err := crud.DB.Exec(`UPDATE items SET note = NULL WHERE name = ?`, "n"); err != nil {
		t.Fatal(err)
	}
	isnull, err := crud.Search(ctx, "items", crudian.SearchQuery{
		Where: crudian.Where().IsNull("note"),
		Limit: 10,
	})
	if err != nil || isnull.Total < 1 {
		t.Fatalf("isNull: %+v %v", isnull, err)
	}

	cnt, err := crud.Count(ctx, "items", crudian.CountQuery{})
	if err != nil || cnt < 3 {
		t.Fatalf("count: %d %v", cnt, err)
	}
	cntW, err := crud.Count(ctx, "items", crudian.CountQuery{Where: crudian.Where().Eq("name", "a")})
	if err != nil || cntW != 1 {
		t.Fatalf("count where: %d %v", cntW, err)
	}

	if _, err := crud.Search(ctx, "items", crudian.SearchQuery{Limit: -1}); err == nil {
		t.Fatal("bad limit")
	}
	if _, err := crud.Search(ctx, "items", crudian.SearchQuery{Where: crudian.Where().In("name", []any{}), Limit: 10}); err == nil {
		t.Fatal("empty in")
	}
}

func TestExtendedWrites(t *testing.T) {
	ctx := context.Background()
	crud := mustCreateCrud(t, openDB(t))

	ins, err := crud.Upsert(ctx, "items", crudian.Row{"id": int64(10), "name": "new", "score": 1})
	if err != nil || toInt64(ins["id"]) != 10 {
		t.Fatalf("upsert insert: %+v %v", ins, err)
	}
	upd, err := crud.Upsert(ctx, "items", crudian.Row{"id": int64(10), "name": "upd", "score": 9})
	if err != nil || upd["name"] != "upd" || toInt64(upd["score"]) != 9 {
		t.Fatalf("upsert update: %+v %v", upd, err)
	}
	if _, err := crud.Upsert(ctx, "items", crudian.Row{"name": "x", "score": 1}); err == nil {
		t.Fatal("upsert without id")
	}

	src, err := crud.Create(ctx, "items", crudian.Row{"name": "src", "score": 5})
	if err != nil {
		t.Fatal(err)
	}
	dup, err := crud.Duplicate(ctx, "items", crudian.DuplicateQuery{
		Where:     crudian.Where().Eq("id", src["id"]),
		Overrides: crudian.Row{"name": "dup"},
	})
	if err != nil || dup["name"] != "dup" || toInt64(dup["id"]) == toInt64(src["id"]) {
		t.Fatalf("dup: %+v %v", dup, err)
	}
	miss, err := crud.Duplicate(ctx, "items", crudian.DuplicateQuery{Where: crudian.Where().Eq("id", 99999)})
	if err != nil || miss != nil {
		t.Fatalf("dup miss: %+v %v", miss, err)
	}

	n, err := crud.BulkCreate(ctx, "items", []crudian.Row{
		{"name": "b1", "score": 1},
		{"name": "b2", "score": 2},
	})
	if err != nil || n != 2 {
		t.Fatalf("bulkCreate: %d %v", n, err)
	}
	if z, err := crud.BulkCreate(ctx, "items", nil); err != nil || z != 0 {
		t.Fatalf("bulkCreate empty: %d %v", z, err)
	}

	un, err := crud.BulkUpdate(ctx, "items", crudian.Row{"score": 99}, crudian.UpdateQuery{
		Where: crudian.Where().Gte("score", 1).And(crudian.Where().Lte("score", 2)),
	})
	if err != nil || un < 1 {
		t.Fatalf("bulkUpdate: %d %v", un, err)
	}

	bn, err := crud.BulkUpsert(ctx, "items", []crudian.Row{
		{"id": int64(20), "name": "u1", "score": 1},
		{"id": int64(21), "name": "u2", "score": 2},
	})
	if err != nil || bn != 2 {
		t.Fatalf("bulkUpsert: %d %v", bn, err)
	}
	if _, err := crud.BulkUpsert(ctx, "items", []crudian.Row{{"name": "no-id", "score": 1}}); err == nil {
		t.Fatal("bulkUpsert without id")
	}

	dn, err := crud.BulkDelete(ctx, "items", crudian.DeleteQuery{Where: crudian.Where().Gte("id", 20)})
	if err != nil || dn < 1 {
		t.Fatalf("bulkDelete: %d %v", dn, err)
	}
	if _, err := crud.BulkUpdate(ctx, "items", crudian.Row{"score": 1}, crudian.UpdateQuery{}); err == nil {
		t.Fatal("bulkUpdate needs where")
	}
}

func TestTransaction(t *testing.T) {
	ctx := context.Background()
	crud := mustCreateCrud(t, openDB(t))

	err := crud.Transaction(ctx, func(tx *crudian.Crud) error {
		if _, err := tx.Create(ctx, "items", crudian.Row{"name": "t1", "score": 1}); err != nil {
			return err
		}
		if _, err := tx.Create(ctx, "items", crudian.Row{"name": "t2", "score": 2}); err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	cnt, err := crud.Count(ctx, "items", crudian.CountQuery{})
	if err != nil || cnt != 2 {
		t.Fatalf("committed: %d %v", cnt, err)
	}

	err = crud.Transaction(ctx, func(tx *crudian.Crud) error {
		if _, err := tx.Create(ctx, "items", crudian.Row{"name": "boom", "score": 3}); err != nil {
			return err
		}
		return crudian.NewError("rollback")
	})
	if err == nil {
		t.Fatal("expected rollback error")
	}
	cnt2, err := crud.Count(ctx, "items", crudian.CountQuery{})
	if err != nil || cnt2 != 2 {
		t.Fatalf("rolled back: %d %v", cnt2, err)
	}
}

func toInt64(v any) int64 {
	switch n := v.(type) {
	case int64:
		return n
	case int:
		return int64(n)
	case int32:
		return int64(n)
	case float64:
		return int64(n)
	default:
		return 0
	}
}

File: packages/js/src/libsql/crud.test.ts
import assert from "node:assert/strict"
import { mkdtempSync, rmSync } from "node:fs"
import { tmpdir } from "node:os"
import { join } from "node:path"
import { after, afterEach, before, describe, test } from "node:test"
import { createClient, type Client } from "@libsql/client"
import { CrudianError, where } from "../index.js"
import { createCrud } from "./index.js"

const SCHEMA = `
  CREATE TABLE IF NOT EXISTS items (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT NOT NULL,
    score INTEGER,
    note TEXT
  );
`

let tempDir = ""
let client: Client

async function resetTable(c: Client) {
  await c.execute("DELETE FROM items")
  try {
    await c.execute("DELETE FROM sqlite_sequence WHERE name = 'items'")
  } catch {
    // sqlite_sequence may be absent until the first AUTOINCREMENT insert
  }
}

async function crudFixture() {
  await resetTable(client)
  return createCrud(client)
}

before(async () => {
  tempDir = mkdtempSync(join(tmpdir(), "crudian-libsql-"))
  const dbPath = join(tempDir, "test.db")
  client = createClient({ url: `file:${dbPath}` })
  await client.execute(SCHEMA)
})

afterEach(async () => {
  await resetTable(client)
})

after(() => {
  client.close()
  if (tempDir) rmSync(tempDir, { recursive: true, force: true })
})

describe("libsql.createCrud", () => {
  test("正常系: 注入クライアントと同一参照", async () => {
    const crud = await crudFixture()
    assert.equal(crud.db, client)
  })

  test("正常系: メソッド一式を持つ", async () => {
    const crud = await crudFixture()
    for (const name of [
      "create",
      "read",
      "update",
      "delete",
      "upsert",
      "duplicate",
      "bulkCreate",
      "bulkUpdate",
      "bulkDelete",
      "bulkUpsert",
      "search",
      "list",
      "count",
      "transaction",
    ] as const) {
      assert.equal(typeof crud[name], "function")
    }
  })

  test("異常系: null / undefined / 想定外型は拒否", () => {
    assert.throws(() => createCrud(null as never), CrudianError)
    assert.throws(() => createCrud(undefined as never), CrudianError)
    assert.throws(() => createCrud("not-db" as never), CrudianError)
  })
})

describe("libsql.transaction", () => {
  test("正常系: コミットと戻り値", async () => {
    const crud = await crudFixture()
    const result = await crud.transaction(async () => {
      await crud.create("items", { name: "a", score: 1 })
      await crud.create("items", { name: "b", score: 2 })
      return "ok"
    })
    assert.equal(result, "ok")
    assert.equal((await crud.search("items")).items.length, 2)
  })

  test("異常系: throw でロールバック", async () => {
    const crud = await crudFixture()
    await assert.rejects(
      () =>
        crud.transaction(async () => {
          await crud.create("items", { name: "a", score: 1 })
          throw new Error("boom")
        }),
      /boom/,
    )
    assert.equal((await crud.search("items")).items.length, 0)
  })

  test("異常系: 非関数は拒否", async () => {
    const crud = await crudFixture()
    await assert.rejects(() => crud.transaction("nope" as never), CrudianError)
  })
})

describe("libsql.coreCrud", () => {
  test("正常系: create / read / update / delete", async () => {
    const crud = await crudFixture()
    const created = await crud.create("items", { name: "alpha", score: 10, note: "n" })
    assert.equal(created.id, 1)
    assert.deepEqual(
      await crud.read("items", {
        columns: ["id", "name"],
        where: where().eq("id", 1),
      }),
      { id: 1, name: "alpha" },
    )
    const updated = await crud.update(
      "items",
      { score: 9 },
      { where: where().eq("id", 1) },
    )
    assert.equal(updated?.score, 9)
    assert.equal(updated?.note, "n")
    assert.equal(await crud.delete("items", { where: where().eq("id", 1) }), 1)
  })

  test("異常系: 未ヒット・0件・検証", async () => {
    const crud = await crudFixture()
    assert.equal(await crud.read("items", { where: where().eq("id", 999) }), null)
    assert.equal(
      await crud.update("items", { score: 1 }, { where: where().eq("id", 999) }),
      null,
    )
    assert.equal(await crud.delete("items", { where: where().eq("id", 999) }), 0)
    await assert.rejects(() => crud.create(1 as never, { name: "a" }), CrudianError)
    await assert.rejects(() => crud.create("items", { score: 1 }))
  })
})

describe("libsql.count", () => {
  test("正常系: 空表 / 全件 / where", async () => {
    const crud = await crudFixture()
    assert.equal(await crud.count("items"), 0)
    await crud.create("items", { name: "a", score: 1 })
    await crud.create("items", { name: "b", score: 2 })
    assert.equal(await crud.count("items"), 2)
    assert.equal(await crud.count("items", { where: where().eq("name", "a") }), 1)
  })

  test("異常系: テーブル名 / in 空", async () => {
    const crud = await crudFixture()
    await assert.rejects(() => crud.count(1 as never), CrudianError)
    await assert.rejects(
      () => crud.count("items", { where: where().in("name", []) }),
      CrudianError,
    )
  })
})

describe("libsql.searchList", () => {
  test("正常系: ページングと list 別名", async () => {
    const crud = await crudFixture()
    for (let i = 0; i < 5; i++) {
      await crud.create("items", { name: `n${i}`, score: i })
    }
    const page1 = await crud.search("items", { limit: 2 })
    assert.equal(page1.items.length, 2)
    assert.equal(page1.hasMore, true)
    assert.equal(page1.nextCursor, 2)
    assert.equal(page1.total, 5)
    const page2 = await crud.search("items", { limit: 2, cursor: page1.nextCursor })
    assert.deepEqual(
      page2.items.map((i) => i.id),
      [3, 4],
    )
    assert.equal(page2.total, 5)
    const q = { limit: 10, where: where().eq("name", "n0") }
    assert.deepEqual(await crud.list("items", q), await crud.search("items", q))
  })

  test("正常系: where 演算子", async () => {
    const crud = await crudFixture()
    await crud.create("items", { name: "alice", score: 10, note: null })
    await crud.create("items", { name: "bob", score: 20, note: "x" })
    await crud.create("items", { name: "carol", score: 30, note: "y" })
    assert.equal(
      (await crud.search("items", { where: where().in("name", ["alice", "bob"]) }))
        .items.length,
      2,
    )
    assert.equal(
      (await crud.search("items", { where: where().like("name", "a%") })).items[0]
        ?.name,
      "alice",
    )
    assert.equal(
      (await crud.search("items", { where: where().isNull("note") })).items.length,
      1,
    )
  })

  test("異常系: limit / in 空", async () => {
    const crud = await crudFixture()
    await assert.rejects(() => crud.search("items", { limit: 0 }), CrudianError)
    await assert.rejects(
      () => crud.search("items", { where: where().in("name", []) }),
      CrudianError,
    )
  })
})

describe("libsql.extendedWrites", () => {
  test("正常系: upsert / duplicate / bulk*", async () => {
    const crud = await crudFixture()
    const inserted = await crud.upsert("items", { id: 10, name: "new", score: 1 })
    assert.equal(inserted.id, 10)
    await crud.upsert("items", { id: 10, name: "upd", score: 9 })
    assert.equal(
      (await crud.read("items", { where: where().eq("id", 10) }))?.name,
      "upd",
    )

    const source = await crud.create("items", { name: "a", score: 1, note: "n" })
    const copy = await crud.duplicate("items", {
      where: where().eq("id", source.id),
      overrides: { name: "b" },
    })
    assert.notEqual(copy?.id, source.id)
    assert.equal(copy?.name, "b")
    assert.equal(
      await crud.duplicate("items", { where: where().eq("id", 999) }),
      null,
    )

    assert.equal(
      await crud.bulkCreate("items", [
        { name: "x", score: 1 },
        { name: "y", score: 2 },
      ]),
      2,
    )
    assert.equal(await crud.bulkCreate("items", []), 0)
    assert.equal(
      await crud.bulkUpdate("items", { score: 8 }, { where: where().eq("name", "x") }),
      1,
    )
    assert.equal(await crud.bulkDelete("items", { where: where().eq("name", "y") }), 1)
    assert.equal(
      await crud.bulkUpsert("items", [
        { id: 100, name: "z", score: 1 },
        { id: 10, name: "z2", score: 2 },
      ]),
      2,
    )
  })

  test("異常系: id / where 欠落", async () => {
    const crud = await crudFixture()
    await assert.rejects(() => crud.upsert("items", { name: "a" }), CrudianError)
    await assert.rejects(() => crud.duplicate("items", {} as never), CrudianError)
    await assert.rejects(() => crud.bulkUpsert("items", [{ name: "a" }]), CrudianError)
    await assert.rejects(
      () => crud.bulkUpdate("items", { score: 1 }, {} as never),
      CrudianError,
    )
  })
})

describe("libsql.adapterExports", () => {
  test("正常系: createCrud を export する", async () => {
    const mod = await import("./index.js")
    assert.equal(typeof mod.createCrud, "function")
  })
})
